etk-asm/src/asm.rs

```diff
@@ -430,202 +343,163 @@ impl Assembler {
+
+                        self.undeclared_labels.extend(labels);
+                        self.ready.push(rop.clone());
                     }
+                    Err(ops::Error::ContextIncomplete {
+                        source: UnknownMacro { name, .. },
+                    }) => return error::UndeclaredInstructionMacro { name }.fail(),
+                    Err(ops::Error::ContextIncomplete {
+                        source: UndefinedVariable { name, .. },
+                    }) => return error::UndeclaredVariableMacro { var: name }.fail(),
                 }
-
-                Ok(())
             }
             RawOp::Raw(raw) => {
                 self.concrete_len += raw.len();
-                self.ready.extend(raw);
-                Ok(())
+                self.ready.push(RawOp::Raw(raw.to_vec()));
             }
-        }
-    }
-
-    fn push_pending(&mut self, rop: RawOp) -> Result<(), Error> {
-        // Update total size of pending ops.
-        if let Some(ref mut pending_len) = self.pending_len {
-            match rop.size() {
-                Some(size) => *pending_len += size,
-                None => self.pending_len = None,
+            RawOp::Scope(scope) => {
+                let mut asm = Self::new();
+                let scope_result = asm.assemble(&scope)?;
+                self.concrete_len += scope_result.len();
+                self.ready.push(RawOp::Raw(scope_result));
             }
         }
 
-        // Handle new label and macro definitions.
-        match (self.pending_len, rop) {
-            (Some(pending_len), RawOp::Op(AbstractOp::Label(lbl))) => {
-                // The label has a defined address.
-                let address = self.concrete_len + pending_len;
-                let item = self.declared_labels.get_mut(&*lbl).unwrap();
-                *item = Some(address);
-            }
-            (None, rop @ RawOp::Op(AbstractOp::Label(_))) => {
-                self.pending.push_back(rop);
-                if self.undeclared_labels.is_empty() {
-                    self.choose_sizes()?;
-                }
-            }
-            (_, RawOp::Op(AbstractOp::MacroDefinition(defn))) => {
-                if let Some(RawOp::Op(AbstractOp::Macro(invc))) = self.pending.front() {
-                    if defn.name() == &invc.name {
-                        let invc = invc.clone();
-                        self.pending.pop_front();
-                        self.expand_macro(&invc.name, &invc.parameters)?;
+        Ok(self.concrete_len)
+    }
+
+    fn backpatch_labels(&mut self) -> Result<(), Error> {
+        for op in self.variable_sized_push.iter() {
+            if let AbstractOp::Push(imm) = op {
+                let exp = imm
+                    .tree
+                    .eval_with_context((&self.declared_labels, &self.declared_macros).into());
+
+                if let Ok(val) = exp {
+                    let imm_size = (val.to_f64().unwrap().log2() / 8.0).ceil() as usize;
```

Comment:
Can we use [`BigInt::bits`](https://docs.rs/num-bigint/latest/num_bigint/struct.BigInt.html#method.bits) and [integer ceiling division](https://stackoverflow.com/a/2745086) here instead of converting to a float? I'm not sure if this'll give the correct answer for really big numbers.